tests/testapps/AddressBookSample/AddressBook/SingleContactViewController.m

```diff
@@ -0,0 +1,417 @@
+        // remove it entirely if the user cleared the text field.
+        if (phoneNumber) {
+            succeeded = succeeded && ABMultiValueReplaceValueAtIndex(phoneNumbers, phoneNumber, 0);
+        } else {
+            succeeded = succeeded && ABMultiValueRemoveValueAndLabelAtIndex(phoneNumbers, 0);
+        }
+
+        succeeded = succeeded && ABRecordSetValue(self.contact, kABPersonPhoneProperty, phoneNumbers, &error);
+        CFRelease(phoneNumbers);
+    }
+
+    if (phoneNumber) {
+        CFRelease(phoneNumber);
+    }
+
+    if (self.email.isEnabled) {
+        ABMultiValueRef existingEmail = ABRecordCopyValue(self.contact, kABPersonEmailProperty);
+        ABMutableMultiValueRef emails = ABMultiValueCreateMutableCopy(existingEmail);
+        CFRelease(existingEmail);
+
+        // Either update the value of the first email address (since that's the one we displayed), or
+        // remove it entirely if the user cleared the text field.
+        if (email) {
+            succeeded = succeeded && ABMultiValueReplaceValueAtIndex(emails, email, 0);
+        } else {
+            succeeded = succeeded && ABMultiValueRemoveValueAndLabelAtIndex(emails, 0);
+        }
+
+        succeeded = succeeded && ABRecordSetValue(self.contact, kABPersonEmailProperty, emails, &error);
+        CFRelease(emails);
+    }
+
+    if (email) {
+        CFRelease(email);
+    }
+
+    succeeded = succeeded && ABAddressBookSave(self.addressBook, &error);
+    if (!succeeded) {
+        NSLog(@"There was an error modifying the contact: %@", (__bridge_transfer NSString*)CFErrorCopyDescription(error));
+    }
+}
+
+- (void)_createContact {
+    CFErrorRef error = NULL;
+    self.contact = ABPersonCreate();
+    CFTypeRef firstName = [self _copyContactInfoFrom:self.firstName];
+    CFTypeRef lastName = [self _copyContactInfoFrom:self.lastName];
+    CFTypeRef phoneNumber = [self _copyContactInfoFrom:self.phoneNumber];
+    CFTypeRef email = [self _copyContactInfoFrom:self.email];
+    CFTypeRef address = [self _copyContactInfoFrom:self.address];
+    bool succeeded = true;
+
+    if (firstName) {
+        succeeded = succeeded && ABRecordSetValue(self.contact, kABPersonFirstNameProperty, firstName, &error);
+        CFRelease(firstName);
+    }
+
+    if (lastName) {
+        succeeded = succeeded && ABRecordSetValue(self.contact, kABPersonLastNameProperty, lastName, &error);
+        CFRelease(lastName);
+    }
+
+    if (phoneNumber) {
+        ABMutableMultiValueRef phoneNumbers = ABMultiValueCreateMutable(kABMultiStringPropertyType);
+        succeeded = succeeded && ABMultiValueAddValueAndLabel(phoneNumbers, phoneNumber, kABPersonPhoneMainLabel, NULL);
+        succeeded = succeeded && ABRecordSetValue(self.contact, kABPersonPhoneProperty, phoneNumbers, &error);
+        CFRelease(phoneNumbers);
+        CFRelease(phoneNumber);
+    }
+
+    if (email) {
```

Comment:
nit: I would prefer validity checks like this when the objects are assigned/created, that way you can easily exit on failure cases and don't have to have multiple checks for validity of memory throughout the code. Note: you can wrap them in CFAutoRelease, so when you exit on failure cases you don't have to have explicit releases for the ones above. If not you can write a macro to release multiple objects. Although autorelease seems more appropriate.